### ISSUE TITLE
`kotlinx.coroutines.debug`'s `DebugProbes` fails when used from a different ClassLoader as `kotlin.stdlib`

### ISSUE THREAD

**Issue by ArchangelX360** [Kotlin/kotlinx.coroutines#4000]
**Describe the bug**

`DebugProbes.install()` works, but the code crashed when spawning the first coroutine after installation, with `java.lang.ClassNotFoundException: kotlinx.coroutines.debug.internal.DebugProbesImpl`.

First, a quick context on our different module layers hierarchy (we use JPMS), our module layers are structured like such:
```
Layer "fleet.dock.bootstrap" (`AppClassLoader`)
    -> Module `fleet.dock.bootstrap`
    -> Module `kotlin.stdlib`
    -> [...]
  └── Layer "fleet.dock.api"
      -> Module `fleet.dock.api`
      -> Module `kotlinx.coroutines.core`
      -> Module `kotlinx.coroutines.debug`
      -> [...]
```
_Note: Module layers does not have names, I just name them by the most significant module that is part of it for convenience._

Here is what I think is happening, internally, when doing `DebugProbes.install()`, `kotlinx.coroutines.debug` uses ByteBuddy to [swap](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-debug/src/internal/Attach.kt#L20-L30) `kotlin.stdlib`'s `kotlin.coroutines.jvm.internal.DebugProbesKt` code with the [code from `kotlinx.coroutines.core`'s `kotlinx.coroutines.debug.internal.DebugProbesKt`](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbes.kt).
This runtime injected code tries to use [`DebugProbesImpl` object](https://github.com/Kotlin/kotlinx.coroutines/blob/39ea109008e7de1233f7b650fd88980f037b5c1b/kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbes.kt#L18C62-L18C77) from `kotlinx.coroutines.core` module. But it does that at the `kotlin.stdlib` level, so using the `fleet.dock.bootstrap`'s ClassLoader (`AppClassLoader`), which does not know about `kotlinx.coroutines.core` module and so it crashes with `ClassNotFoundException`.

**Comment by qwwdfsad** [comment]
Thanks! `DebugProbes`, indeed, were not designed/implemented to work with multiple classloaders and with module separation in mind.

Could you please elaborate on the following:

* Is it correct that this problem prevents you from using debug probes in Fleet? Could you please share FL ticket?
* Would you be able to test an intermediate -dev build w.r.t. this issue? Classloading is a nuanced matter and it would be nice to have an extra check before shipping the fix

**Comment by ArchangelX360** [comment]
> Is it correct that this problem prevents you from using debug probes in Fleet? Could you please share FL ticket?

Indeed, we solve the package split internally, but we are stuck on that one and there is nothing we can do on our side. Here is the ticket: https://youtrack.jetbrains.com/issue/FL-23308

> Would you be able to test an intermediate -dev build w.r.t. this issue? Classloading is a nuanced matter and it would be nice to have an extra check before shipping the fix

Yes, that's most probably possible, and we would be our pleasure to help testing it out.

**Comment by qwwdfsad** [comment]
Thanks!

**Comment by qwwdfsad** [comment]
We've discussed it further and I also investigated why IDEA does not have this problem.

The short summary:

* There is nothing we can do on `kotlinx-coroutines` side. Indeed, `DebugProbes` from stdlib should be replaced with something and this something is located in `kotlinx-coroutines` and depends on it. We cannot reasonably isolate the debugger from `kotlinx-coroutines`, thus, this is the intrinsic requirement of the debugger -- to be in the same CL hierarchy/JPMS reachability as the standard library. This is the reason `docs` tag is here -- we'll explicitly mention that
* IntelliJ right now loads both stdlib and coroutines in the same classloader, but previously used child-first classloader (see `com.intellij.ide.plugins.cl.PluginClassLoader`) and, when requested to load a class by a classloader holding standard library, first looked the class up in children classloader (not unconditionally, but with specific filtering)

If you or anybody else have any other ideas how we can patch these things up, I'm open to suggestions
